new_lineage_panel/src/ActionWidget.tsx

```diff
@@ -0,0 +1,392 @@
+import {
+  Dispatch,
+  FC,
+  FunctionComponent,
+  PropsWithChildren,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import "reactflow/dist/style.css";
+import { Button, Card, CardBody, Input, Label, Tooltip } from "reactstrap";
+import AlertCircleIcon from "./assets/icons/alert-circle.svg?react";
+import ResetIcon from "./assets/icons/reset.svg?react";
+import HelpIcon from "./assets/icons/help.svg?react";
+import FeedbackIcon from "./assets/icons/feedback.svg?react";
+import ExpandIcon from "./assets/icons/expand.svg?react";
+import ArrowLeftDoubleIcon from "./assets/icons/arrow-left-double.svg?react";
+import ArrowRightDoubleIcon from "./assets/icons/arrow-right-double.svg?react";
+import ArrowLeftIcon from "./assets/icons/arrow-left.svg?react";
+import ArrowRightIcon from "./assets/icons/arrow-right.svg?react";
+import styles from "./styles.module.scss";
+import { HELP_SIDEBAR } from "./constants";
+import { init, openURL, setLegacyLineageView, CLL } from "./service_utils";
+import { LineageContext, aiEnabled } from "./App";
+import { useReactFlow } from "reactflow";
+import {
+  calculateMinLevel,
+  expandTableLineageLevelWise,
+  highlightTableConnections,
+  layoutElementsOnCanvas,
+} from "./graph";
+import classNames from "classnames";
+import { BetterPopover } from "./components/Modal/BetterPopover";
+
+const InfoIcon: FunctionComponent<{ id: string; message: string }> = ({
+  id,
+  message,
+}) => {
+  const [isOpen, setIsOpen] = useState(false);
+  return (
+    <div
+      className={styles.alert_icon}
+      id={id}
+      onMouseEnter={() => setIsOpen(true)}
+      onMouseLeave={() => setIsOpen(false)}
+    >
+      <AlertCircleIcon />
+      <Tooltip target={id} isOpen={isOpen}>
+        {message}
+      </Tooltip>
+    </div>
+  );
+};
+
+const ActionButton = ({
+  onClick,
+  children,
+}: PropsWithChildren<{
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+}>) => {
+  return (
+    <Button
+      size="sm"
+      outline
+      color="secondary"
+      onClick={onClick}
+      className="d-flex align-items-center gap-sm"
+    >
+      {children}
+    </Button>
+  );
+};
+
+const AutoExpansionPopover = () => {
+  const flow = useReactFlow();
+  const {
+    selectedTable,
+    leftExpansion,
+    setLeftExpansion,
+    rightExpansion,
+    setRightExpansion,
+    minRange,
+    setMinRange,
+    rerender,
+  } = useContext(LineageContext);
+  const [nodeCount, setNodeCount] = useState(0);
+  const [maxRange, setMaxRange] = useState([0, 0]);
+
+  useEffect(() => {
+    setLeftExpansion(Math.max(minRange[0], 0));
+    setRightExpansion(Math.max(minRange[1], 0));
+  }, [minRange, setLeftExpansion, setRightExpansion]);
+
+  useEffect(() => {
+    (async () => {
```

Comment:
it is better to avoid this pattern of self invoking async functions inside useEffect to avoid repeatedly calling the async function. pls extract into useCallback function and invoke it here